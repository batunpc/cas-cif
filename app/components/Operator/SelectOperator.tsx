import React, { useState } from "react";
import type { SelectOperator_query$key } from "__generated__/SelectOperator_query.graphql";
import { graphql, useFragment } from "react-relay";
import { Dropdown } from "semantic-ui-react";
import safeJsonParse from "lib/safeJsonParse";

interface Props {
  query: SelectOperator_query$key;
  applyChange: (x: any) => void;
  formChangeData: any;
}

const SelectOperator: React.FC<Props> = ({
  query,
  applyChange,
  formChangeData,
}) => {
  const queryData = useFragment(
    graphql`
      fragment SelectOperator_query on Query {
        query {
          allOperators {
            edges {
              node {
                id
                rowId
                legalName
                tradeName
                bcRegistryId
              }
            }
          }
        }
      }
    `,
    query
  );

  const selectedOperator = formChangeData.operator_id
    ? queryData.query.allOperators.edges.find(
        (edge) => edge.node.rowId === formChangeData.operator_id
      ).node
    : null;

  const handleChange = (e, data) => {
    e.preventDefault();
    e.stopPropagation();
    applyChange({ operator_id: safeJsonParse(data.value).rowId });
  };

  return (
    <>
      <fieldset>
        <legend>Identity</legend>
        <strong>Legal Operator Name and BC Registry ID</strong>
        <Dropdown
          id="operator-dropdown"
          placeholder="Select an Operator"
          fluid
          search
          selection
          defaultValue={JSON.stringify(selectedOperator)}
          onChange={(e, data) => {
            handleChange(e, data);
          }}
          options={queryData.query.allOperators.edges.map(({ node }) => {
            return {
              key: node.id,
              text: `${node.legalName} (${node.bcRegistryId})`,
              value: JSON.stringify(node),
            };
          })}
        />
        <strong>Trade Name</strong>
        {selectedOperator?.tradeName}
      </fieldset>
      <style jsx>
        {`
          strong {
            font-size: 1.2em;
            margin-bottom: 1em;
            margin-top: 1em;
          }
          fieldset {
            padding: 2em;
            border: 2px solid silver;
          }
          ,
          legend {
            color: silver;
            padding: 2px;
            align: center;
            display: block;
            text-align: center;
            font-size: 1.2em;
          }
          ,
          #operator-dropdown {
            text-align: left;
            color: black;
          }
        `}
      </style>
    </>
  );
};

export default SelectOperator;
