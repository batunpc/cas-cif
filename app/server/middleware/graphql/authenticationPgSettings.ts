import { getPriorityGroup } from "../../../lib/userGroups";
import { getUserGroups } from "../../helpers/userGroupAuthentication";
import groupData from "../../../data/groups.json";
import { isAuthenticated } from "@bcgov-cas/sso-express";
import type { Request } from "express";

const AUTH_BYPASS_COOKIE = "mocks.auth";

const AS_CIF_INTERNAL = process.argv.includes("AS_CIF_INTERNAL");

const AS_CIF_EXTERNAL = process.argv.includes("AS_CIF_EXTERNAL");

const AS_CIF_ADMIN = process.argv.includes("AS_CIF_ADMIN");

const AS_UNAUTHORIZED_IDIR = process.argv.includes("AS_UNAUTHORIZED_IDIR");
const AS_CYPRESS = process.argv.includes("AS_CYPRESS");

const allowCypressForRole = (roleName: string, req: Request) => {
  return AS_CYPRESS && req.cookies[AUTH_BYPASS_COOKIE] === roleName;
};

const authenticationPgSettings = (req: Request) => {
  if (AS_CIF_INTERNAL || allowCypressForRole("cif_internal", req)) {
    return {
      "jwt.claims.sub": "00000000-0000-0000-0000-000000000000",
      "jwt.claims.user_groups": "cif_internal",
      "jwt.claims.priority_group": "cif_internal",
      role: "cif_internal",
    };
  }

  if (AS_CIF_EXTERNAL || allowCypressForRole("cif_external", req)) {
    return {
      "jwt.claims.sub": "00000000-0000-0000-0000-000000000001",
      "jwt.claims.user_groups": "cif_external",
      "jwt.claims.priority_group": "cif_external",
      role: "cif_external",
    };
  }

  if (AS_CIF_ADMIN || allowCypressForRole("cif_admin", req)) {
    return {
      "jwt.claims.sub": "00000000-0000-0000-0000-000000000002",
      "jwt.claims.user_groups": "cif_admin",
      "jwt.claims.priority_group": "cif_admin",
      role: "cif_admin",
    };
  }

  if (AS_UNAUTHORIZED_IDIR || allowCypressForRole("unauthorized_idir", req)) {
    return {
      "jwt.claims.sub": "00000000-0000-0000-0000-000000000000",
      "jwt.claims.user_groups": "UNAUTHORIZED_IDIR",
      "jwt.claims.priority_group": "UNAUTHORIZED_IDIR",
      role: "ciip_guest",
    };
  }

  const groups = getUserGroups(req);
  const priorityGroup = getPriorityGroup(groups);

  const claimsSettings = {
    role: groupData[priorityGroup].pgRole,
  };
  if (!isAuthenticated(req))
    return {
      ...claimsSettings,
    };

  const claims = req.claims;

  claims.user_groups = groups.join(",");
  claims.priority_group = priorityGroup;

  const properties = [
    "jti",
    "exp",
    "nbf",
    "iat",
    "iss",
    "aud",
    "sub",
    "typ",
    "azp",
    "auth_time",
    "session_state",
    "acr",
    "email_verified",
    "name",
    "preferred_username",
    "given_name",
    "family_name",
    "email",
    "broker_session_id",
    "user_groups",
    "priority_group",
  ];
  properties.forEach((property) => {
    claimsSettings[`jwt.claims.${property}`] = claims[property];
  });

  return {
    ...claimsSettings,
  };
};

export default authenticationPgSettings;
