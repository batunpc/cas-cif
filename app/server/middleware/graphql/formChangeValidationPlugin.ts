import { makeWrapResolversPlugin } from "postgraphile";
import validationSchemas from "../../../data/jsonSchemaForm/validationSchemas";
import Ajv, { ErrorObject } from "ajv";

const ajv = new Ajv({ allErrors: true });

function validateRecord(schemaName: string, formData: any): ErrorObject[] {
  const schema = validationSchemas[schemaName];

  // ajv caches compiled schemas on first instantiation, we don't need to
  // precompile schemas in advance
  const validate = ajv.compile(schema);
  const valid = validate(formData);

  return valid ? [] : validate.errors;
}

export const FormChangeValidationPlugin = makeWrapResolversPlugin(
  (context) => {
    if (
      context.scope.isRootMutation &&
      context.scope.fieldName === "updateFormChange"
    ) {
      return {
        // There is no need to pass a context to the resolver at this time
      };
    }
    return null;
  },
  () => async (resolver: any, source, args, context, resolveInfo) => {
    // If the mutation doesn't change the form data, we don't need to re-validate
    if (args.input?.formChangePatch?.newFormData === undefined)
      return resolver();

    const {
      identifiers: [formChangeId],
    } = (resolveInfo as any).graphile.build.getTypeAndIdentifiersFromNodeId(
      args.input.id
    );

    // This is the recommended way to fetch data ahead of running the
    // resolver of the GraphQL mutation - see first "IMPORTANT" note here:
    // https://www.graphile.org/postgraphile/make-extend-schema-plugin/#the-selectgraphqlresultfromtable-helper
    const { pgClient } = context;
    const {
      rows: [formChangeRecord],
    } = await pgClient.query(
      `select json_schema_name from cif.form_change where id = $1`,
      [formChangeId]
    );

    const errors = validateRecord(
      formChangeRecord.json_schema_name,
      args.input.formChangePatch.newFormData
    );

    args.input.formChangePatch = {
      ...args.input.formChangePatch,
      validationErrors: errors,
    };

    const result = await resolver(source, args, context, resolveInfo);
    return result;
  }
);
