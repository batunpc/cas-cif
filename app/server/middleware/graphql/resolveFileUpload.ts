import config from "../../../config";

async function saveRemoteFile({ stream }) {
  const response = await fetch(
    `${config.get("storageApiHost")}/api/v1/attachments/upload`,
    {
      method: "POST",
      headers: {
        "api-key": config.get("storageApiKey"),
        "Content-Type": "multipart/form-data",
      },
      body: stream,
    }
  );
  try {
    return await response.json();
  } catch (e) {
    console.error(e);
  }
}

export default async function resolveFileUpload(upload) {
  const { createReadStream } = upload;
  const stream = createReadStream();

  // Save tile to remote storage system
  const { uuid } = await saveRemoteFile({ stream });

  return uuid;
}
