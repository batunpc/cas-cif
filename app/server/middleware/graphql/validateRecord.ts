import Ajv, { ErrorObject } from "ajv";
import { KEY_EMAIL, KEY_PHONE_NO, KEY_RFP_DIGITS, REGEX_EMAIL, REGEX_PHONE_NO, REGEX_RFP_DIGITS } from "data/validation-constants";

const ajv = new Ajv({ allErrors: true });
// AJV needs to be made aware of any custom formats used in the schema
ajv.addFormat(KEY_RFP_DIGITS, REGEX_RFP_DIGITS);
ajv.addFormat(KEY_EMAIL, REGEX_EMAIL);
ajv.addFormat(KEY_PHONE_NO, REGEX_PHONE_NO);

const validateRecord: (schema: any, formData: any) => ErrorObject[] = (
  schema,
  formData
) => {
  // ajv caches compiled schemas on first instantiation, we don't need to
  // precompile schemas in advance
  const validate = ajv.compile(schema);
  const valid = validate(formData);

  return valid ? [] : validate.errors;
};

export default validateRecord;
