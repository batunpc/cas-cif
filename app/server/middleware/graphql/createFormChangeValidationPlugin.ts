import { GraphQLObjectType, GraphQLResolveInfo } from "graphql";
import { Context, makeWrapResolversPlugin } from "postgraphile";
import validateRecord from "./validateRecord";
import databaseSchemaConfiguration from "./databaseSchemaConfiguration";
import validationSchemas from "../../../data/jsonSchemaForm/validationSchemas";

const DATABASE_SCHEMAS = databaseSchemaConfiguration;

export const filter = (context: Context<GraphQLObjectType<any, any>>) => {
  if (
    context.scope.isRootMutation &&
    context.scope.fieldName === "createFormChange"
  ) {
    return {
      // There is no need to pass a context to the resolver at this time
    };
  }
  return null;
};

export const resolverWrapperGenerator =
  () =>
  async (
    resolver: any,
    source,
    args: { [argName: string]: any },
    context,
    resolveInfo: GraphQLResolveInfo
  ) => {
    if (!args.input.jsonSchemaName) {
      throw new Error("JSON schema must be set in the mutation input");
    }
    const { pgClient } = context;

    let jsonSchema;

    if (DATABASE_SCHEMAS.includes(args.input.jsonSchemaName)) {
      const {
        rows: [fetchedRecord],
      } = await pgClient.query(
        `select json_schema from cif.form
         where slug = $1`,
        [args.input.jsonSchemaName]
      );
      jsonSchema = fetchedRecord.json_schema.schema;
    } else jsonSchema = validationSchemas[args.input.jsonSchemaName];

    const errors = validateRecord(jsonSchema, args.input.newFormData ?? {});

    args.input.validationErrors = errors;

    const result = await resolver(source, args, context, resolveInfo);
    return result;
  };

const CreateFormChangeValidationPlugin = makeWrapResolversPlugin(
  filter,
  resolverWrapperGenerator
);

export default CreateFormChangeValidationPlugin;
