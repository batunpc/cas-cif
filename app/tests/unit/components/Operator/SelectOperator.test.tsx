import { render, screen } from "@testing-library/react";
import SelectOperator from "components/Operator/SelectOperator";
import { mockRandom } from "jest-mock-random";
import { SelectOperator_query } from "__generated__/SelectOperator_query.graphql";

beforeEach(() => {
  mockRandom([0.3, 0.1, 0.4, 0.1, 0.5, 0.9]);
});

describe("The SelectOperator component", () => {
  it("matches the snapshot", () => {
    const queryData: SelectOperator_query = {
      query: {
        allOperators: {
          edges: [
            {
              node: {
                id: "abc",
                rowId: 1,
                legalName: "Test Operator LEGAL",
                tradeName: "Test Operator TRADE",
                bcRegistryId: "123456789",
              },
            },
            {
              node: {
                id: "def",
                rowId: 2,
                legalName: "Second Test Operator LEGAL",
                tradeName: "Second Test Operator TRADE",
                bcRegistryId: "987654321",
              },
            },
          ],
        },
      },
      " $refType": "SelectOperator_query",
    };

    jest
      .spyOn(require("react-relay"), "useFragment")
      .mockImplementation(() => queryData);

    const componentUnderTest = render(
      <SelectOperator
        query={null}
        applyChange={() => null}
        formChangeData={{}}
      />
    );

    expect(componentUnderTest.container).toMatchSnapshot();
  });

  it("renders n options for n edges in allOperators", () => {
    const queryData: SelectOperator_query = {
      query: {
        allOperators: {
          edges: [
            {
              node: {
                id: "abc",
                rowId: 1,
                legalName: "Test Operator LEGAL",
                tradeName: "Test Operator TRADE",
                bcRegistryId: "123456789",
              },
            },
            {
              node: {
                id: "def",
                rowId: 2,
                legalName: "Second Test Operator LEGAL",
                tradeName: "Second Test Operator TRADE",
                bcRegistryId: "987654321",
              },
            },
          ],
        },
      },
      " $refType": "SelectOperator_query",
    };

    jest
      .spyOn(require("react-relay"), "useFragment")
      .mockImplementation(() => queryData);

    render(
      <SelectOperator
        query={null}
        applyChange={() => null}
        formChangeData={{}}
      />
    );

    expect(screen.getAllByRole("option").length).toEqual(2);
  });

  it("initializes the selected operator when one exists in the formChangeData", () => {
    const queryData: SelectOperator_query = {
      query: {
        allOperators: {
          edges: [
            {
              node: {
                id: "abc",
                rowId: 1,
                legalName: "Test Operator LEGAL",
                tradeName: "Test Operator TRADE",
                bcRegistryId: "123456789",
              },
            },
            {
              node: {
                id: "def",
                rowId: 2,
                legalName: "Second Test Operator LEGAL",
                tradeName: "Second Test Operator TRADE",
                bcRegistryId: "987654321",
              },
            },
          ],
        },
      },
      " $refType": "SelectOperator_query",
    };

    jest
      .spyOn(require("react-relay"), "useFragment")
      .mockImplementation(() => queryData);

    render(
      <SelectOperator
        query={null}
        applyChange={() => null}
        formChangeData={{ operator_id: 1 }}
      />
    );

    expect(screen.getByRole("alert").textContent).toBe(
      "Test Operator LEGAL (123456789)"
    );
  });
});
