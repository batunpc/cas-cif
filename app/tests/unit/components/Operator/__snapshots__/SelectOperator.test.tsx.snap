// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The SelectOperator component matches the snapshot 1`] = `
<div>
  <fieldset
    class="jsx-355091279"
  >
    <legend
      class="jsx-355091279"
    >
      Identity
    </legend>
    <strong
      class="jsx-355091279"
    >
      Legal Operator Name and BC Registry ID
    </strong>
    <div
      aria-expanded="false"
      class="ui fluid search selection dropdown"
      id="operator-dropdown"
      role="combobox"
    >
      <input
        aria-autocomplete="list"
        autocomplete="off"
        class="search"
        tabindex="0"
        type="text"
        value=""
      />
      <div
        aria-atomic="true"
        aria-live="polite"
        class="divider text"
        role="alert"
      >
        Select an Operator
      </div>
      <i
        aria-hidden="true"
        class="dropdown icon"
      />
      <div
        class="menu transition"
        role="listbox"
      >
        <div
          aria-checked="false"
          aria-selected="true"
          class="selected item"
          role="option"
          style="pointer-events: all;"
        >
          <span
            class="text"
          >
            Test Operator LEGAL (123456789)
          </span>
        </div>
        <div
          aria-checked="false"
          aria-selected="false"
          class="item"
          role="option"
          style="pointer-events: all;"
        >
          <span
            class="text"
          >
            Second Test Operator LEGAL (987654321)
          </span>
        </div>
      </div>
    </div>
    <strong
      class="jsx-355091279"
    >
      Trade Name
    </strong>
  </fieldset>
</div>
`;
